extras/kms/kms.go

```diff
@@ -204,6 +216,155 @@ func (k *Kms) GetWrapper(ctx context.Context, scopeId string, purpose KeyPurpose
 	return wrapper, nil
 }
 
+// CreateKeys creates the root key and DEKs for the given scope id.  By
+// default, CreateKeys manages its own transaction (begin/rollback/commit).
+//
+// It's valid to provide no KeyPurposes (nil or empty), which means that the
+// scope will only end up with a root key (and one rk version) with no DEKs.
+//
+// CreateKeys also supports the WithTx(...) and WithReaderWriter(...) options
+// which allows the caller to pass an inflight transaction to be used for all
+// database operations.  If WithTx(...) or WithReaderWriter(...) are used, then
+// the caller is responsible for managing the transaction.  The purpose of the
+// WithTx(...) and WithReaderWriter(...) options are to allow the caller to
+// create the scope and all of its keys in the same transaction.
+//
+// The WithRandomReadear(...) option is supported as well.  If no optional
+// random reader is provided, then the reader from crypto/rand will be used as
+// a default.
+func (k *Kms) CreateKeys(ctx context.Context, scopeId string, purposes []KeyPurpose, opt ...Option) error {
+	const op = "kms.(Kms).CreateKeysTx"
+	if scopeId == "" {
+		return fmt.Errorf("%s: missing scope id: %w", op, ErrInvalidParameter)
+	}
+	for _, p := range purposes {
+		if !purposeListContains(k.purposes, p) {
+			return fmt.Errorf("%s: not a supported key purpose %q: %w", op, p, ErrInvalidParameter)
+		}
+	}
+
+	rootWrapper, err := k.GetExternalRootWrapper()
+	if err != nil {
+		return fmt.Errorf("%s: %w", op, err)
+	}
+
+	opts := getOpts(opt...)
+
+	if isNil(opts.withRandomReader) {
+		opts.withRandomReader = rand.Reader
+	}
+
+	var localTx *dbw.RW
+	var r dbw.Reader
+	var w dbw.Writer
+	switch {
+	case opts.withTx != nil:
+		if opts.withReader != nil || opts.withWriter != nil {
+			return fmt.Errorf("%s: WithTx(...) and WithReaderWriter(...) options cannot be used at the same time: %w", op, ErrInvalidParameter)
+		}
+		if ok := opts.withTx.IsTx(); !ok {
+			return fmt.Errorf("%s: provided transaction has no inflight transaction: %w", op, ErrInvalidParameter)
+		}
+		r = opts.withTx
+		w = opts.withTx
+	case opts.withReader != nil, opts.withWriter != nil:
+		if opts.withTx != nil {
+			return fmt.Errorf("%s: WithTx(...) and WithReaderWriter(...) options cannot be used at the same time: %w", op, ErrInvalidParameter)
+		}
+		if opts.withReader == nil {
+			return fmt.Errorf("%s: WithReaderWriter(...) option is missing the reader: %w", op, ErrInvalidParameter)
+		}
+		if opts.withWriter == nil {
+			return fmt.Errorf("%s: WithReaderWriter(...) option is missing the writer: %w", op, ErrInvalidParameter)
+		}
+		r = opts.withReader
+		w = opts.withWriter
+	default:
+		localTx, err = k.repo.writer.Begin(ctx)
+		if err != nil {
+			return fmt.Errorf("%s: unable to begin transaction: %w", op, err)
+		}
+		r = localTx
+		w = localTx
```

Comment:
Instead of creating a `localTx` here and then checking for it, what do you think about deferring the operations we want to run only for local transactions?
```suggestion
		localTx, err = k.repo.writer.Begin(ctx)
		if err != nil {
			return fmt.Errorf("%s: unable to begin transaction: %w", op, err)
		}
		defer func() {
			if retErr == nil {
				if err  := localTx.Commit(ctx); err != nil {
					retErr = fmt.Errorf("%s: unable to commit transaction: %w", op, err)
				}
			} else {
				if err := localTx.Rollback(ctx); err != nil {
					retErr = multierror.Append(retErr, err)
				}
			}
		}()
		r = localTx
		w = localTx
```

Then we can remove the `localTx` checks and operations below. It would require us to use a named returned (name `retErr` here) to allow us to mutate the error in the defer.

Reply:
ty for the suggestion... played around a bit and decided to leave it basically as-is for now.